go/vt/wrangler/vexec_plan.go

```diff
@@ -0,0 +1,261 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package wrangler
+
+import (
+	"fmt"
+
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+type vexecPlan struct {
+	query       string
+	opcode      int
+	parsedQuery *sqlparser.ParsedQuery
+}
+
+const (
+	insertQuery = iota
+	updateQuery
+	deleteQuery
+	selectQuery
+)
+
+func (vx *vexec) buildVExecPlan() (*vexecPlan, error) {
+	stmt, err := sqlparser.Parse(vx.query)
+	if err != nil {
+		return nil, err
+	}
+	var plan *vexecPlan
+	switch stmt := stmt.(type) {
+	case *sqlparser.Insert:
+		plan, err = vx.buildInsertPlan(stmt)
+	case *sqlparser.Update:
+		plan, err = vx.buildUpdatePlan(stmt)
+	case *sqlparser.Delete:
+		plan, err = vx.buildDeletePlan(stmt)
+	case *sqlparser.Select:
+		plan, err = vx.buildSelectPlan(stmt)
+	default:
+		return nil, fmt.Errorf("query not supported by vexec: %s", sqlparser.String(stmt))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	plan.query = vx.query
+	vx.plan = plan
+	return plan, nil
+}
+
+func splitAndExpression(filters []sqlparser.Expr, node sqlparser.Expr) []sqlparser.Expr {
+	if node == nil {
+		return filters
+	}
+	switch node := node.(type) {
+	case *sqlparser.AndExpr:
+		filters = splitAndExpression(filters, node.Left)
+		return splitAndExpression(filters, node.Right)
+	}
+	return append(filters, node)
+}
+
+func (vx *vexec) analyzeWhere(where *sqlparser.Where) []string {
+	var cols []string
+	if where == nil {
+		return cols
+	}
+	exprs := splitAndExpression(nil, where.Expr)
+	for _, expr := range exprs {
+		switch expr := expr.(type) {
+		case *sqlparser.ComparisonExpr:
+			qualifiedName, ok := expr.Left.(*sqlparser.ColName)
+			if ok {
+				cols = append(cols, qualifiedName.Name.String())
+			}
+		}
+	}
+	return cols
+}
+
+func (vx *vexec) addDefaultWheres(where *sqlparser.Where) *sqlparser.Where {
+	cols := vx.analyzeWhere(where)
+	var hasDBName, hasWorkflow bool
+	for _, col := range cols {
+		if col == "db_name" {
+			hasDBName = true
+		} else if col == "workflow" {
+			hasWorkflow = true
+		}
+	}
+	newWhere := where
+	if !hasDBName {
+		expr := &sqlparser.ComparisonExpr{
+			Left:     &sqlparser.ColName{Name: sqlparser.NewColIdent("db_name")},
+			Operator: sqlparser.EqualStr,
+			Right:    sqlparser.NewStrVal([]byte(vx.masters[0].DbName())),
+		}
+		if newWhere == nil {
+			newWhere = &sqlparser.Where{
+				Type: sqlparser.WhereStr,
+				Expr: expr,
+			}
+		} else {
+			newWhere.Expr = &sqlparser.AndExpr{
+				Left:  newWhere.Expr,
+				Right: expr,
+			}
+		}
+	}
+	if !hasWorkflow {
+		expr := &sqlparser.ComparisonExpr{
+			Left:     &sqlparser.ColName{Name: sqlparser.NewColIdent("workflow")},
+			Operator: sqlparser.EqualStr,
+			Right:    sqlparser.NewStrVal([]byte(vx.workflow)),
+		}
+		newWhere.Expr = &sqlparser.AndExpr{
+			Left:  newWhere.Expr,
+			Right: expr,
+		}
+	}
+	return newWhere
+}
+
+func (vx *vexec) buildInsertPlan(ins *sqlparser.Insert) (*vexecPlan, error) {
```

Comment:
This is something we have to watch, and may have to remove. I can't think of a use case where this will make sense yet. Commands like `Reshard` etc. are better abstractions in some sense.

Reply:
I removed this.

I had not implemented it at first, but we decided to add this for completeness. But yes, it looks pretty useless atm: we can implement any special workflows when we see a need.